File: python/shared/dtos/auction_dto.py
from enum import IntEnum, StrEnum

from . import BaseDTO, BigInt


class AuctionStatus(IntEnum):
    WAITING = 0
    IN_PROGRESS = 1
    COMPLETED = 2


class MessageType(StrEnum):
    AUTH = "auth"
    TIMER = "timer"
    PLACE_BID = "place_bid"
    BID_PLACED = "bid_placed"
    MEMBER_SOLD = "member_sold"
    MEMBER_UNSOLD = "member_unsold"
    NEXT_MEMBER = "next_member"
    QUEUE_UPDATE = "queue_update"
    INIT = "init"
    STATUS = "status"
    ERROR = "error"
    USER_CONNECTED = "user_connected"
    USER_DISCONNECTED = "user_disconnected"


class Team(BaseDTO):
    team_id: int
    leader_id: int
    member_id_list: list[int]
    points: int


class AuctionStateDTO(BaseDTO):
    auction_id: BigInt
    status: AuctionStatus
    current_member_id: int | None
    current_bid: int | None
    current_bidder: int | None
    timer: int
    teams: list[Team]
    auction_queue: list[int]
    unsold_queue: list[int]
    connected_users: list[int]
    preset_snapshot: dict | None


class AuctionDTO(BaseDTO):
    auction_id: BigInt


class CreateAuctionDTO(BaseDTO):
    is_public: bool
    send_invite: bool


class TimerMessageData(BaseDTO):
    timer: int


class StatusMessageData(BaseDTO):
    status: str


class NextMemberMessageData(BaseDTO):
    member_id: int


class QueueUpdateMessageData(BaseDTO):
    auction_queue: list[int]
    unsold_queue: list[int]


class MemberSoldMessageData(BaseDTO):
    teams: list[Team]


class BidPlacedMessageData(BaseDTO):
    team_id: int
    leader_id: int
    amount: int


class WebSocketMessage(BaseDTO):
    type: str
    data: dict

File: web/src/dtos/auctionDto.ts
export enum AuctionStatus {
  WAITING = 0,
  IN_PROGRESS = 1,
  COMPLETED = 2,
}

export interface AuctionDTO {
  auctionId: string;
}

export type MessageType =
  | "auth"
  | "timer"
  | "place_bid"
  | "bid_placed"
  | "member_sold"
  | "member_unsold"
  | "next_member"
  | "queue_update"
  | "init"
  | "status"
  | "error"
  | "user_connected"
  | "user_disconnected";

export interface WebSocketMessage {
  type: MessageType;
  data: any;
}

export interface Team {
  teamId: number;
  leaderId: number;
  memberIdList: number[];
  points: number;
}

export interface StatusMessageData {
  status: string;
}

export interface AuctionInitDTO {
  auctionId: string;
  status: AuctionStatus;
  currentMemberId: number | null;
  currentBid: number | null;
  currentBidder: number | null;
  timer: number;
  teams: Team[];
  auctionQueue: number[];
  unsoldQueue: number[];
  teamId: number | null;
  memberId: number | null;
  isLeader: boolean;
  connectedUsers: number[];
  presetSnapshot: any | null;
}

export interface BidPlacedMessageData {
  teamId: number;
  leaderId: number;
  amount: number;
}

export interface NextMemberMessageData {
  memberId: number;
}

export interface QueueUpdateMessageData {
  auctionQueue: number[];
  unsoldQueue: number[];
}

export interface MemberSoldMessageData {
  teams: Team[];
}

export interface TimerMessageData {
  timer: number;
}
